Support for Coq "vos" builds

rlepigre opened the pull request:
This is an alternative to https://github.com/ocaml/dune/pull/4050, implementing @ejgallego's idea from https://github.com/ocaml/dune/pull/4050#issuecomment-1300916807.

The "vos" mode is enabled using `(mode vos)` in `coq.theory` stanzas.

CC @Alizter.

Commit:
[coq] support for vos builds.

This is configured using a new [vos] mode, that can be configured
in the [coq.theory] stanza.

Signed-off-by: Rodolphe Lepigre <rodolphe@bedrocksystems.com>

Alizter commented on src/dune_rules/coq/coq_rules.ml:
```diff
@@ -132,6 +132,16 @@ let coqc_native_flags ~sctx ~dir ~theories_deps ~theory_dirs
       ]
+  | VosOnly ->
+    Command.Args.As
+      [ "-vos"
```
Since you add vos in file_flags do you need this here? Can be the same as VoOnly I think.

rlepigre replied:
Yeah, the other one (variable `file_flags`) is for `coqdep`.

rlepigre replied:
I opted to always call `coqdep` with the `-vos` option, so that we can share the dependency analysis between modes.

Alizter commented:
What happens when a user builds once in `VoOnly` stops the build realising they should have `VosOnly` and then continues? Doesn't this make the build inconsistent? IIRC Coq loads vo files if they are available and then vos files in vos mode.

rlepigre commented:
Since you configure `vo` or `vos` per theory, you can't have both `vo` and `vos` at the same time. Will that not save us?

Alizter commented:
The issue is that Coq goes and finds the files itself so if you have a dirty build directory you might corrupt the build.

Alizter commented:
What I am suggesting is the opposite situation. If you had 3 files:
```
A.v -> B.v -> C.v
```
Now if I had `(mode vo)` and built `B.vo` I would have `A.vo` and `B.vo` in my build directory. Now if I switch to `(mode vos)` I will and build `C.vos`, Dune will give us: `A.vos`, `B.vos` and `C.vos`. However when compiling `C.vos`, `coqc` would have used `B.vo` rather than `B.vos`.

Alizter commented:
Actually nevermind. Reading the Coq refman I see that `vos` is indeed preferred over `vo`. https://coq.inria.fr/refman/practical-tools/coq-commands.html#compiled-interfaces-produced-using-vos

rlepigre commented on bin/coq/coqtop.ml:
```diff
@@ -93,17 +93,24 @@ let term =
               , extr.buildable.coq_lang_version
-              , "DuneExtraction" )
+              , "DuneExtraction"
+              , extr.buildable.mode )
```
No idea, I never actually used extraction. But I don't see why it shouldn't, and it felt like it would make sense to use the information since we have it.

rlepigre replied:
Should we do anything here @Alizter?

Alizter replied:
No nothing actionable, just a question.

ejgallego reviewed (changes requested):
It is a pity we can't be more dynamic yet but the patch is a start thanks!

Needs documentation / changes entry, rest looks good to me!

Alizter commented:
One more design consideration is if we want users to have to edit their stanzas at all. Perhaps it would be OK to have both rules at the same time as they are independent. We could even add `vos` to the `check` alias for example.

ejgallego commented:
> What happens when a user builds once in `VoOnly` stops the build realising they should have `VosOnly` and then continues? Doesn't this make the build inconsistent? IIRC Coq loads vo files if they are available and then vos files in vos mode.

You don't have that problem as all the `.vo` targets in the first run become stale (as rules won't generate them anymore) so Dune will clean them up.

ejgallego commented:
> @ejgallego It won't become nondeterministic see my retort above.

If I understood your comment above, you actually are giving an example of a non-determinitic build for C.vos, as it now depends on what was built before.

Alizter commented:
No, according to the Coq refman, vos is always preferred in -vos mode.

ejgallego commented:
Tho indeed maybe the race is for the vo objects, let me check coq's source

rlepigre commented:
One question I have is: can we somehow make it so that we can configure the mode in the profile? Is it already possible with what I did?

Alizter commented:
@rlepigre We can detect the `-vos` flag in `(coq (flags ...))` if a user has that and prefer the mode that way. I think it is better to do that in a later PR though.

If I am correct about being able to have both the vo and vos rules at the same time, then that would mean we wouldn't need it anyway.

Blaisorblade commented:
@Alizter coqc _can_ mix vo and vos files, so you do need separate directories; also the `vo` build overrides `vos`/`vok` files but dune might be able to handle _that_. And since we need separate profiles, can we use that to enable `vos` mode?

Let me [quote the manual](https://coq.inria.fr/refman/practical-tools/coq-commands.html#compiled-interfaces-produced-using-vos) (since there's no anchor to link)

> **Interaction with standard compilation**
>
> When compiling a file foo.v using coqc in the standard way (i.e., without -vos nor -vok), an empty file foo.vos and an empty file foo.vok are created in addition to the regular output file foo.vo. If coqc is subsequently invoked on some other file bar.v using option -vos or -vok, and that bar.v requires foo.v, if Coq finds an empty file foo.vos, then it will load foo.vo instead of foo.vos.
>
> The purpose of this feature is to allow users to benefit from the -vos option even if they depend on libraries that were compiled in the traditional manner (i.e., never compiled using the -vos option).

ejgallego commented:
Indeed I'm not sure we have the right mechanism to perform this kind of selection at runtime, it seems to me that using flags is not better than using `mode`, as both require updating the stanza which we def don't want.

I can't check the exact dependency rules now, so indeed the next step would be to see if we can enable both modes without getting unsoundness as Ali proposes.

In the immediate term, I propose we merge this PR once the docs are added, and continue the discussion in the issue / subsequent PR

rlepigre commented:
I guess I need to change the Coq language version @ejgallego?

Alizter commented:
@rlepigre You don't need to bump the coq lang version since for Dune 3.8 it will be 0.8 and we haven't released Dune yet.

ejgallego commented:
> @rlepigre You don't need to bump the coq lang version since for Dune 3.8 it will be 0.8 and we haven't released Dune yet.

Indeed, thanks @Alizter ; the changelog of language versions should make this clear by maybe adding the Dune version next to the coq lang version https://github.com/ocaml/dune/blob/main/doc/coq.rst#coq-language-version 

Note that indeed we bumped the lang in the implementationhttps://github.com/ocaml/dune/blob/main/src/dune_rules/coq/coq_stanza.ml#L4
but not in the docs.

Alizter commented:
It will be bumped in the docs, that is what https://github.com/ocaml/dune/pull/7384 is about.

rlepigre commented:
Sorry, I'm just only seeing your last messages now: I pushed some documentation, but maybe not exactly what you had in mind.

Alizter commented on doc/coq.rst:
```diff
@@ -152,6 +152,12 @@ The semantics of the fields are:
   Previous versions of Dune before 3.7 would disable the native rules depending
   on whether or not the ``dev`` profile was selected.
 
+- Form version :ref:`Coq lang 0.8<coq-lang>` onwards, ``(mode vos)`` makes it
```
```suggestion
- From version :ref:`Coq lang 0.8<coq-lang>` onwards, ``(mode vos)`` makes it
```

Alizter reviewed (approved):
LGTM I'll let @ejgallego decide when to merge.

rlepigre commented:
Actually, we have a CI failure on the test I added: apparently, the output is not consistent with what I get locally. Any clue why this happens @Alizter? I'm running Coq 8.16.0, and CI seems to be using 8.16.1: could that be the reason?
```diff
diff --git a/_build/.sandbox/d3bcb07409ea29304d4499ea0e5fe788/default/test/blackbox-tests/test-cases/coq/vos-build.t/run.t b/_build/.sandbox/d3bcb07409ea29304d4499ea0e5fe788/default/test/blackbox-tests/test-cases/coq/vos-build.t/run.t.corrected
index 1b5e703..65ef517 100644
--- a/_build/.sandbox/d3bcb07409ea29304d4499ea0e5fe788/default/test/blackbox-tests/test-cases/coq/vos-build.t/run.t
+++ b/_build/.sandbox/d3bcb07409ea29304d4499ea0e5fe788/default/test/blackbox-tests/test-cases/coq/vos-build.t/run.t.corrected
@@ -35,8 +35,8 @@ Checking that we can go back to vo mode (without cleaning).
 
   $ mv dune-vo dune
   $ dune build --display short --debug-dependency-path @all
-          coqc foo.{glob,vo}
-          coqc bar.{glob,vo}
+          coqc Nbasic_foo.{cmi,cmxs},foo.{glob,vo}
+          coqc Nbasic_bar.{cmi,cmxs},bar.{glob,vo}
   $ dune build --debug-dependency-path @default
   lib: [
     "_build/install/default/lib/base/META"
@@ -44,6 +44,10 @@ Checking that we can go back to vo mode (without cleaning).
     "_build/install/default/lib/base/opam"
   ]
   lib_root: [
+    "_build/install/default/lib/coq/user-contrib/basic/.coq-native/Nbasic_bar.cmi" {"coq/user-contrib/basic/.coq-native/Nbasic_bar.cmi"}
+    "_build/install/default/lib/coq/user-contrib/basic/.coq-native/Nbasic_bar.cmxs" {"coq/user-contrib/basic/.coq-native/Nbasic_bar.cmxs"}
+    "_build/install/default/lib/coq/user-contrib/basic/.coq-native/Nbasic_foo.cmi" {"coq/user-contrib/basic/.coq-native/Nbasic_foo.cmi"}
+    "_build/install/default/lib/coq/user-contrib/basic/.coq-native/Nbasic_foo.cmxs" {"coq/user-contrib/basic/.coq-native/Nbasic_foo.cmxs"}
     "_build/install/default/lib/coq/user-contrib/basic/bar.v" {"coq/user-contrib/basic/bar.v"}
     "_build/install/default/lib/coq/user-contrib/basic/bar.vo" {"coq/user-contrib/basic/bar.vo"}
     "_build/install/default/lib/coq/user-contrib/basic/foo.v" {"coq/user-contrib/basic/foo.v"}
```

ejgallego commented:
@rlepigre the difference you see is because dune's CI test with the coq-native opam package installed

Alizter commented on test/blackbox-tests/test-cases/coq/vos-build.t/dune-vo:
```diff
@@ -0,0 +1,9 @@
+(coq.theory
+ (name basic)
```
Add (mode vo) here

Alizter replied:
Dune will autodetect coq-native which is enabled in the test suite. You can force the mode by giving the field

rlepigre replied:
OK thanks. Done.

ejgallego commented:
We can still improve the docs of the `(mode ...)` field but IMHO that can be done in a different PR.
